google/resource_compute_attached_disk.go

```diff
@@ -0,0 +1,231 @@
+			State: resourceAttachedDiskImport,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(300 * time.Second),
+			Delete: schema.DefaultTimeout(240 * time.Second),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"attached_disk": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
+			},
+			"attached_instance": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
+			},
+			"project": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"zone": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"self_link": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAttachedDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	attachedInstance := d.Get("attached_instance").(string)
+	zone, err := getZoneForAttachedDisk(d, config, attachedInstance)
+	if err != nil {
+		return err
+	}
+
+	instanceName := GetResourceNameFromSelfLink(attachedInstance)
+	diskName := GetResourceNameFromSelfLink(d.Get("attached_disk").(string))
+
+	attachedDisk := compute.AttachedDisk{
+		DeviceName: diskName,
+		Source:     fmt.Sprintf("/projects/%s/zones/%s/disks/%s", project, zone, diskName),
+	}
+
+	op, err := config.clientCompute.Instances.AttachDisk(project, zone, instanceName, &attachedDisk).Do()
+	if err != nil {
+		return err
+	}
+
+	// TODO (chrisst) change format of the internal id to include project/zone
+	d.SetId(fmt.Sprintf("%s:%s", instanceName, diskName))
+
+	// TODO (chrisst) allow for override to timeouts
+	waitErr := computeSharedOperationWaitTime(config.clientCompute, op, project, 2, "disk to attach")
+	if waitErr != nil {
+		d.SetId("")
+		return waitErr
+	}
+
+	return resourceAttachedDiskRead(d, meta)
+}
+
+func resourceAttachedDiskRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+	d.Set("project", project)
+
+	attachedInstance := d.Get("attached_instance").(string)
+	zone, err := getZoneForAttachedDisk(d, config, attachedInstance)
+	if err != nil {
+		return err
+	}
+	d.Set("zone", zone)
+
+	instanceName := GetResourceNameFromSelfLink(attachedInstance)
+	diskName := GetResourceNameFromSelfLink(d.Get("attached_disk").(string))
+
+	instance, err := config.clientCompute.Instances.Get(project, zone, instanceName).Do()
+	if err != nil {
+		return err
+	}
+
+	// Force the referenced resources to a self-link in state because it's more specific then name.
+	d.Set("attached_instance", instance.SelfLink)
+	d.Set("attached_disk", fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/disks/%s", project, zone, diskName))
```

Comment:
You *should* be able to get the self link from instance.Disks? In theory the `Source` attribute is a self link though I  could be very wrong.

Reply:
It's actually an [`AttachedDisk`](https://godoc.org/google.golang.org/api/compute/v1#AttachedDisk) object and not a disk instance. From what I can see it would require a 2nd GET in order to retrieve the actual Disk object.